python/ray/dag/compiled_dag_node.py

```diff
@@ -1075,6 +1124,221 @@ def _get_or_compile(
         self._dag_submitter.start()
         self._dag_output_fetcher.start()
 
+    def _build_execution_schedule(self):
+        """
+        Generate an execution schedule for each actor. The schedule is a list of
+        DAGNodeOperation.
+
+        Step 1: Generate a graph based on the following rules:
+
+        #1  Divide a DAG node into three GraphNodes: READ, COMPUTE, and WRITE. Each
+            GraphNode has a DAGNodeOperation.
+        #2  Add edges between READ and COMPUTE, and between COMPUTE and WRITE, which
+            belong to the same task.
+        #3  Add an edge between COMPUTE with bind_index i and COMPUTE with bind_index
+            i+1 if they belong to the same actor.
+        #4  Add an edge between WRITE of the writer task and READ of the reader task.
+
+        Step 2: Topological sort
+
+        If there are multiple GraphNodes with zero in-degree, select one based on
+        the following rules:
+
+        #1  If the nodes are not NCCL write nodes, select the one with the smallest
+            `bind_index`. If there are multiple candidate nodes with the smallest
+            `bind_index` of the actors that they belong to, any one of them is
+            acceptable. For the implementation details, we maintain a priority queue
+            for each actor, where the peek of the priority queue is the node with the
+            smallest `bind_index`.
+        #2  If the node is an NCCL write node, select it only if all of its downstream
+            nodes are also the peeks of their priority queues.
+        #3  If #1 and #2 cannot be satisfied, it means that all candidate nodes are
+            NCCL write nodes. In this case, select the one that is the peek of the
+            priority queue and its downstream nodes, regardless of whether the
+            downstream nodes are peeks of their priority queues or not.
```

Comment:
I think we need to have more explanation of rules 2 and 3, either here or in a ray docs page.

Reply:
Update https://github.com/ray-project/ray/pull/46911/commits/69fa0381069e0af8f5e25ca6c17c068189cb84c5